### DISCUSSION TITLE
Advice on antsMultivariateTemplateConstruction2

### DISCUSSION THREAD

**Discussion by mikearcaro** [ANTsX/ANTs#1609] [Q&A]
Hi!

I'm trying to create a T2 template of 12 subjects. I've run antsMultivariateTemplateConstruction2.sh a couple times (iterating on the prior output) and the template0.nii.gz output doesn't look ideal - there are multiple "shadows" of each individual T2 indicating that the gross shape of the brains are not aligning well. I'm wondering if I'm running the function sub-optimally or maybe not running enough iterations. When I run antsRegistrationSyN.sh to co-register these individual images, they are spot-on. My subjects are rat-sized brains so I'm not dealing with high variability in folding.

For the template construction, the T2s are skull-stripped, N4 bias field corrected brains that have been rigidly aligned (if I don't do rigid alignment then the output template is severely clipped). 
This is what I ran:
antsMultivariateTemplateConstruction2.sh -d 3 -j 4 -i 4 -n 0 -k 1 -f 4x2x1 -r 1 -s 2x1x0vox -q 30x20x4 -t BSplineSyN  -m CC -c 0 -o ShrewTemplate *corrected_brain.nii.gz

I get fantastic registration for any two of my animals if I just run: antsRegistrationSyN.sh -f shrew1.nii.gz -m shrew2.nii.gz -d 3 -o shrew2_to_shrew1 -n 4

Also, if I run antsMultivariateTemplateConstruction2.sh where I define a starting template as the average of all my T2s that have been individually co-registered, then the resulting template is great. Of course, this is not ideal since it will have some degree of bias towards the original reference brain.

Other potentially relevant information: these data are 50um^3.

Any advice would be MUCH appreciated!
Thanks!
Mike

**Comment by ntustison** [comment]
My first guess would be that your transform model is too stiff.  Without explicit specification, the B-spline SyN parameters default to human brain parameters.  Try switching to ``-t SyN`` and see if that works better.

**Comment by cookpa** [comment]
Some users have noticed (with ANTs and other tools) that very small image spacing can affect performance, sometimes these are the result of decisions embedded pretty deep in the underlying algorithms. Temporarily changing the image spacing to be ~ 1mm can help (#1540)

However, if it works well when given an individual reference image, the problem might be a local minimum caused by a low-quality initialization. If initialized from an initial reference image, the template will be biased as you say, but it should converge towards an unbiased solution after many iterations (-i).

Alternatively, you could try to create a better starting average by running a template with a restricted transform. For example, use `-t Affine -A 0 -i 2` - then look at each warped image in the output, look for gross registration failures.

**Comment by gdevenyi** [comment]
For a tool with a number of quality-of-life additions to smooth the rough edges of antsMultivariateTemplateConstruction2.sh, take a look at https://github.com/CoBrALab/optimized_antsMultivariateTemplateConstruction

**Comment by mikearcaro** [comment]
Thank you all for these ideas! 

I tested an affine registration. It actually looks blurrier than what I was previously getting with the template construction. I'll try switching to Syn next.  I do use Syn in antsRegistrationSyN so fingers crossed that works. I'll also take a look at optimized_antsMultivariateTemplateConstruction

**Comment by mikearcaro** [comment]
So using Syn in antsMultivariateTemplateConstruction2 improves things but still results in misalignment in anterior temporal lobe and cerebellum. I can get near spot-on registration with antsRegistrationSyN to a single brain target. Another notable difference is that the average of antsRegistrationSyN is much smoother than the template output from antsMultivariateTemplateConstruction2. Maybe that's a difference in how the brains get averaged vs reflecting alignment differences. To illustrate, I've attached an image showing the results from these alignment.

Does Syn avoid the spacing issues with B-Spline altogether? I'm wondering what other parameters I should try playing in antsMultivariateTemplateConstruction2.

<img width="696" alt="Screen Shot 2023-10-18 at 10 25 44 AM" src="https://github.com/ANTsX/ANTs/assets/27936397/aeed52ad-9eff-4b48-9cda-0da8fe02aac1">

**Reply by cookpa** [reply]
It's certainly possible that the registration parameters differ between the pairwise registration and the template script. The pairwise registration also has a sharper reference image to work with. 

I know @gdevenyi 's scripts have worked well on other data sets, might be worth a try. 

If you can share reproducible examples (code + data) I can take a look, but without that full context it's hard to give much more detailed advice on what's going on here.

**Comment by mikearcaro** [comment]
Yes, I was wondering if this comes down to a sharper reference image to work with. Totally understand it is hard to diagnose with this level of description. The next thing I'll try is the optimized_antsMultivariateTemplateConstruction script. If that fails, I'll put a couple brains in a dropbox folder with the inputs I used to share. I don't want to bother you with that until I've exhausted the options on my end. Thanks!

**Comment by stnava** [comment]
personally -I dont think it's a big deal to start with a single subject as reference.   the scripts "debias" the images anyway, esp after several iterations.  I would just go with that approach ....

**Reply by cookpa** [reply]
Agreed, be sure to turn off rigid initialization (-z <reference image> -r 0) or you'll undo the benefits of using -z.

**Reply by gdevenyi** [reply]
My experience has similarly been that a single subject reference (or an existing average) is a much much better starting target than a rigid average of a population.

**Reply by mydream378** [reply]
@stnava , many thanks for your comments; it saved my life. 
As suggested, I have specified a subject.nii.gz as an initial reference, then the group template is what I NEED.  Please note that I aim to use 60 FA images to created a group template. 

Before reading your comments, it took me almost one week to figure out what is wrong with my over-cut template (see below, Fig. A) via fine-tuning the parameters, e.g., the flags: -m, -t, -s, -c ; all did not work. 

the old codes:

`antsMultivariateTemplateConstruction.sh -d 3 -m 30x50x20 -t GR -s CC -c 2 -j 20  -o FA_final_60_init_053_ *FA.nii.gz`

![2024-12-21 at 18 35](https://github.com/user-attachments/assets/c56f0780-e6b3-4b2d-bc6c-94ccc19818b9)

Inspired by your comments: the new WORKING codes and the derived Fig.B is perfect (`sub-053_ses-1_dti_dtifit_FA.nii.gz` is my specified FA image):

`antsMultivariateTemplateConstruction.sh -d 3 -m 30x50x20 -t GR -s CC -c 2 -j 20 -z sub-053_ses-1_dti_dtifit_FA.nii.gz -y 0 -r 1 -o FA_final_60_init_053_ *FA.nii.gz`

For ones who have the similar issues, as posted before, but without finding the solution.
I hope my comments here will be a NICE stepstone for you: 

https://github.com/ANTsX/ANTs/issues/908

![2024-12-21 at 18 42](https://github.com/user-attachments/assets/b6729cd7-9cdd-4e28-b39f-22ce65f5d59a)

Again, thanks for your genius comments.

**Comment by mikearcaro** [comment]
Just wanted to follow up on this and say that starting with a single good brain as a reference and tweaking the antsMultivariateTemplateConstruction2 parameters to be in line with what I used in antsRegistrationSyN registration worked! The resulting template is crisp. Each subject looks well aligned and the template doesn't look to be biased in shape to the reference. Thanks so much for all of these suggestions!
